### PR TITLE
[FLINK-5026] Rename TimelyFlatMap to Process

### PR DESCRIPTION
**Pull request by aljoscha**
This is only renaming, the changed Javadoc probably requires most attention when reviewing this.

R: @jgrier if you could have a look that would be super

### CODE REVIEW

**Comment by aljoscha** [comment]
R: @kl0u for review, please

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/datastream/ConnectedStreams.java [hunk 1]
```diff
@@ -234,62 +235,62 @@ public StreamExecutionEnvironment getExecutionEnvironment() {
...
 
 	/**
-	 * Applies the given {@link TimelyCoFlatMapFunction} on the connected input streams,
+	 * Applies the given {@link CoProcessFunction} on the connected input streams,
 	 * thereby creating a transformed output stream.
 	 *
 	 * <p>The function will be called for every element in the streams and can produce
 	 * zero or more output. The function can also query the time and set timers. When
 	 * reacting to the firing of set timers the function can emit yet more elements.
 	 *
```
Comment (kl0u) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/datastream/ConnectedStreams.java]: I would transform it into (to also show the distinction between flatMap): 

<p>The function will be called for every element in the input streams and can produce zero or more
	 * output elements. Contrary to the {@link #flatMap(CoFlatMapFunction)} function, this function can also
	 * query the time and set timers. When reacting to the firing of set timers the function can directly
	 * emit elements and/or register yet more timers.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/datastream/ConnectedStreams.java [hunk 2]
```diff
@@ -234,62 +235,62 @@ public StreamExecutionEnvironment getExecutionEnvironment() {
...
+	 * <p>A {@link RichCoProcessFunction}
 	 * can be used to gain access to features provided by the
 	 * {@link org.apache.flink.api.common.functions.RichFunction} interface.
 	 *
-	 * @param coFlatMapper The {@link TimelyCoFlatMapFunction} that is called for each element
+	 * @param coProcessFunction The {@link CoProcessFunction} that is called for each element
 	 *                      in the stream.
 	 *
-	 * @param <R> The of elements emitted by the {@code TimelyCoFlatMapFunction}.
+	 * @param <R> The of elements emitted by the {@code CoProcessFunction}.
```
Comment (kl0u) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/datastream/ConnectedStreams.java]: The type of elements emitted by the {@code CoProcessFunction}.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/datastream/ConnectedStreams.java [hunk 3]
```diff
@@ -234,62 +235,62 @@ public StreamExecutionEnvironment getExecutionEnvironment() {
...
 	 *
```
Comment (kl0u) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/datastream/ConnectedStreams.java]: <R> The type of elements emitted by the {@code CoProcessFunction}.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/datastream/ConnectedStreams.java [hunk 4]
```diff
@@ -234,62 +235,62 @@ public StreamExecutionEnvironment getExecutionEnvironment() {
...
 
 	/**
-	 * Applies the given {@link TimelyCoFlatMapFunction} on the connected input streams,
+	 * Applies the given {@link CoProcessFunction} on the connected input streams,
 	 * thereby creating a transformed output stream.
 	 *
 	 * <p>The function will be called for every element in the streams and can produce
 	 * zero or more output. The function can also query the time and set timers. When
 	 * reacting to the firing of set timers the function can emit yet more elements.
 	 *
```
Comment (kl0u) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/datastream/ConnectedStreams.java]: <p>The function will be called for every element in the input streams and can produce zero or more
	 * output elements. Contrary to the {@link #flatMap(CoFlatMapFunction)} function, this function can also
	 * query the time and set timers. When reacting to the firing of set timers the function can directly
	 * emit elements and/or register yet more timers.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/datastream/KeyedStream.java [hunk 5]
```diff
@@ -173,67 +174,67 @@ public KeyedStream(DataStream<T> dataStream, KeySelector<T, KEY> keySelector, Ty
...
 
 	/**
-	 * Applies the given {@link TimelyFlatMapFunction} on the input stream, thereby
+	 * Applies the given {@link ProcessFunction} on the input stream, thereby
 	 * creating a transformed output stream.
 	 *
 	 * <p>The function will be called for every element in the stream and can produce
 	 * zero or more output. The function can also query the time and set timers. When
 	 * reacting to the firing of set timers the function can emit yet more elements.
 	 *
```
Comment (kl0u) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/datastream/KeyedStream.java]: <p>The function will be called for every element in the input streams and can produce zero or more
	 * output elements. Contrary to the {@link #flatMap(FlatMapFunction)} function, this function can also
	 * query the time and set timers. When reacting to the firing of set timers the function can directly
	 * emit elements and/or register yet more timers.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/datastream/KeyedStream.java [hunk 6]
```diff
@@ -173,67 +174,67 @@ public KeyedStream(DataStream<T> dataStream, KeySelector<T, KEY> keySelector, Ty
...
 	 * can be used to gain access to features provided by the
 	 * {@link org.apache.flink.api.common.functions.RichFunction} interface.
 	 *
-	 * @param flatMapper The {@link TimelyFlatMapFunction} that is called for each element
+	 * @param processFunction The {@link ProcessFunction} that is called for each element
 	 *                      in the stream.
 	 *
-	 * @param <R> The of elements emitted by the {@code TimelyFlatMapFunction}.
+	 * @param <R> The of elements emitted by the {@code ProcessFunction}.
 	 *
```
Comment (kl0u) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/datastream/KeyedStream.java]: <R> The type of elements emitted by the {@code ProcessFunction}.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/datastream/KeyedStream.java [hunk 7]
```diff
@@ -173,67 +174,67 @@ public KeyedStream(DataStream<T> dataStream, KeySelector<T, KEY> keySelector, Ty
...
 
 	/**
-	 * Applies the given {@link TimelyFlatMapFunction} on the input stream, thereby
+	 * Applies the given {@link ProcessFunction} on the input stream, thereby
 	 * creating a transformed output stream.
 	 *
 	 * <p>The function will be called for every element in the stream and can produce
 	 * zero or more output. The function can also query the time and set timers. When
 	 * reacting to the firing of set timers the function can emit yet more elements.
 	 *
```
Comment (kl0u) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/datastream/KeyedStream.java]: <p>The function will be called for every element in the input streams and can produce zero or more
	 * output elements. Contrary to the {@link #flatMap(FlatMapFunction)} function, this function can also
	 * query the time and set timers. When reacting to the firing of set timers the function can directly
	 * emit elements and/or register yet more timers.
	 *

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/datastream/KeyedStream.java [hunk 8]
```diff
@@ -173,67 +174,67 @@ public KeyedStream(DataStream<T> dataStream, KeySelector<T, KEY> keySelector, Ty
...
 	 * {@link org.apache.flink.api.common.functions.RichFunction} interface.
 	 *
-	 * @param flatMapper The {@link TimelyFlatMapFunction} that is called for each element
+	 * @param processFunction The {@link ProcessFunction} that is called for each element
 	 *                      in the stream.
 	 * @param outputType {@link TypeInformation} for the result type of the function.
 	 *
-	 * @param <R> The of elements emitted by the {@code TimelyFlatMapFunction}.
+	 * @param <R> The of elements emitted by the {@code ProcessFunction}.
 	 *
```
Comment (kl0u) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/datastream/KeyedStream.java]: <R> The type of elements emitted by the {@code ProcessFunction}.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/ProcessFunction.java [hunk 9]
```diff
@@ -24,32 +24,24 @@
...
  *
- * <pre>{@code
- * DataStream<X> input = ...;
- *
- * DataStream<Y> result = input.flatMap(new MyTimelyFlatMapFunction());
- * }</pre>
+ * <p>The function will be called for every element in the input stream and can produce
+ * zero or more output. The function can also query the time and set timers. When
+ * reacting to the firing of set timers the function can emit yet more elements.
  *
```
Comment (kl0u) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/ProcessFunction.java]: <p>The function will be called for every element in the input stream and can produce
 * zero or more output elements. Contrary to the {@link org.apache.flink.api.common.functions.FlatMapFunction},
 * this function can also query the time (both event and processing) and set timers,
 * through the provided {@link Context}. When reacting to the firing of set timers the function can
 * directly emit a result, and/or register a timer that will trigger an action in the future.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/co/CoProcessFunction.java [hunk 10]
```diff
@@ -50,7 +46,7 @@
  * @param <OUT> Output type.
```
Comment (kl0u) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/co/CoProcessFunction.java]: I would make it like:

/**
 * A function that processes elements of two streams and produces a single output one.
 *
 * <p>The function will be called for every element in the input streams and can produce
 * zero or more output elements. Contrary to the {@link CoFlatMapFunction}, this function can also
 * query the time (both event and processing) and set timers, through the provided {@link Context}.
 * When reacting to the firing of set timers the function can emit yet more elements.
 * 
 * <p>An example use-case for connected streams would be the application of a set of rules that change
 * over time ({@code stream A}) to the elements contained in another stream (stream {@code B}). The rules
 * contained in {@code stream A} can be stored in the state and wait for new elements to arrive on
 * {@code stream B}. Upon reception of a new element on {@code stream B}, the function can now apply the
 * previously stored rules to the element and directly emit a result, and/or register a timer that
 * will trigger an action in the future.
 
Also we could have a code snippet for the example.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/co/CoProcessFunction.java [hunk 11]
```diff
@@ -50,7 +46,7 @@
...
  */
 @PublicEvolving
-public interface TimelyCoFlatMapFunction<IN1, IN2, OUT> extends Function, Serializable {
+public interface CoProcessFunction<IN1, IN2, OUT> extends Function, Serializable {
 
 	/**
 	 * This method is called for each element in the first of the connected streams.
```
Comment (kl0u) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/co/CoProcessFunction.java]: <p>This function can output zero or more elements using the {@link Collector} parameter
	 * and also update internal state or set timers using the {@link Context} parameter.

Comment (kl0u) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/co/CoProcessFunction.java]: Also in the javadoc for the ctx, it should change to:

A {@link Context} that allows querying the timestamp of...

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/co/CoProcessFunction.java [hunk 12]
```diff
@@ -78,7 +74,7 @@
 	 * @throws Exception The function may throw exceptions which cause the streaming program
 	 *                   to fail and go into recovery.
 	 */
-	void flatMap2(IN2 value, Context ctx, Collector<OUT> out) throws Exception;
+	void processElement2(IN2 value, Context ctx, Collector<OUT> out) throws Exception;
```
Comment (kl0u) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/co/CoProcessFunction.java]: Same comments as for the `processElement1`

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/co/CoProcessFunction.java [hunk 13]
```diff
@@ -119,7 +115,7 @@
 	/**
 	 * Information available in an invocation of {@link #onTimer(long, OnTimerContext, Collector)}.
 	 */
-	interface OnTimerContext extends TimelyFlatMapFunction.Context {
+	interface OnTimerContext extends ProcessFunction.Context {
 		/**
```
Comment (kl0u) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/co/CoProcessFunction.java]: It should extend the local `Context`, not the `ProcessFunction` one.

File: flink-streaming-scala/src/main/scala/org/apache/flink/streaming/api/scala/ConnectedStreams.scala [hunk 14]
```diff
@@ -101,30 +101,31 @@ class ConnectedStreams[IN1, IN2](javaStream: JavaCStream[IN1, IN2]) {
...
 
   /**
-   * Applies the given [[TimelyCoFlatMapFunction]] on the connected input streams,
+   * Applies the given [[CoProcessFunction]] on the connected input streams,
    * thereby creating a transformed output stream.
    *
    * The function will be called for every element in the streams and can produce
    * zero or more output. The function can also query the time and set timers. When
    * reacting to the firing of set timers the function can emit yet more elements.
    *
```
Comment (kl0u) [inline flink-streaming-scala/src/main/scala/org/apache/flink/streaming/api/scala/ConnectedStreams.scala]: The function will be called for every element in the input streams and can produce zero or more
   * output elements. Contrary to the [[ConnectedStreams#flatMap]] function, this function can also
   * query the time and set timers. When reacting to the firing of set timers the function can directly
   * emit elements and/or register yet more timers.

File: flink-streaming-scala/src/main/scala/org/apache/flink/streaming/api/scala/ConnectedStreams.scala [hunk 15]
```diff
@@ -101,30 +101,31 @@ class ConnectedStreams[IN1, IN2](javaStream: JavaCStream[IN1, IN2]) {
...
+   * A [[RichCoProcessFunction]]
    * can be used to gain access to features provided by the
    * [[org.apache.flink.api.common.functions.RichFunction]] interface.
    *
-   * @param coFlatMapper The [[TimelyCoFlatMapFunction]] that is called for each element
-    *                     in the stream.
-    *
+   * @param coProcessFunction The [[CoProcessFunction]] that is called for each element
+    *                    in the stream.
    * @return The transformed { @link DataStream}.
```
Comment (kl0u) [inline flink-streaming-scala/src/main/scala/org/apache/flink/streaming/api/scala/ConnectedStreams.scala]: The transformed [[DataStream]].

File: flink-streaming-scala/src/main/scala/org/apache/flink/streaming/api/scala/KeyedStream.scala [hunk 16]
```diff
@@ -54,28 +54,28 @@ class KeyedStream[T, K](javaStream: KeyedJavaStream[T, K]) extends DataStream[T]
...
 
   /**
-    * Applies the given [[TimelyFlatMapFunction]] on the input stream, thereby
+    * Applies the given [[ProcessFunction]] on the input stream, thereby
     * creating a transformed output stream.
     *
     * The function will be called for every element in the stream and can produce
     * zero or more output. The function can also query the time and set timers. When
     * reacting to the firing of set timers the function can emit yet more elements.
     *
```
Comment (kl0u) [inline flink-streaming-scala/src/main/scala/org/apache/flink/streaming/api/scala/KeyedStream.scala]: The function will be called for every element in the input streams and can produce zero or more
    * output elements. Contrary to the [[KeyedStream#flatMap]] function, this function can also
    * query the time and set timers. When reacting to the firing of set timers the function can directly
    * emit elements and/or register yet more timers.

**Comment by aljoscha** [comment]
Thanks for the thorough review, @kl0u! I'll integrate your comments and merge.

**Comment by aljoscha** [comment]
Manually merged

**Pull request closed without merging**